### DISCUSSION TITLE
Allow storing attachments binary data to another storage service?

### DISCUSSION THREAD

**Discussion by mcrio** [ravendb/ravendb#13673] [Ideas]
Hi. Do you think it would make sense allowing extending the attachments storage mechanism in order to allow saving binary data to another place than the node itself?

Use case:

I am rewriting a system that has lots of attachments. Attachments are not public so accessing them has to go through the application layer and permission checking. Using RavenDb attachment feature would make life much easier. My concern using it at this point is the total size of the attachments and the fact they are stored on all nodes. I fear the attachments might become a bottleneck due to the total size and would induce higher infrastructure costs due to storage requirements on the DB servers. I guess the optimal solution would be extending RavenDb to allow storing the binary at cloud provider storage service, remote machine etc?

**Comment by haludi** [comment]
Hi,

You can save the attachment in a dedicated database.
This database can belong to a database group of one node if you don't want
it to be replicated (but then if the server is down you will not
have access to those attachments).
Then you can configure this specific database to be located on another
mounted volume -
https://ravendb.net/docs/article-page/5.0/csharp/server/storage/customizing-raven-data-files-locations
In the origin database, you can save property with the id of the document
that contains the attachment.
Note, there is no way to send one request (with one transaction) to both
databases so you need to handle errors by yourself.
You can also use external storage and reference it from the document (maybe
by the attachment URL).

Anyway, moving part of the database data to another location is not
practical.

Best regards,
Igal

**Comment by ayende** [comment]
The key issue here is that at this point, they are no longer attachments. They are just links to some other place. At that point, the question is why do you want to have them managed directly by RavenDB? What is the benefit here?

We have this issue: https://issues.hibernatingrhinos.com/issue/RavenDB-16815

That is about retiring attachments over time, not storing them elsewhere in general.

About the size across the cluster issue, note that this depends on how important the data is. You probably wants redundancy for the data, no? But otherwise, you can just upload that to a cloud storage and hold the key instead.

Attachments in RavenDB are actually involved in a lot of things (transactions, indexing, backups, etc). Making them external would create quite a few complicated scenarios.

**Reply by mcrio** [reply]
Thanks to both. I guess I was just after an easier way to manage the attachments and their lifetime instead of implementing my own solution. But yeah, I see your point how making them external complicates currently available features and promises.
